src/main/java/org/cloudfoundry/promregator/lite/discovery/SimpleTargetResolver.kt

```diff
@@ -0,0 +1,191 @@
+package org.cloudfoundry.promregator.lite.discovery
+
+import org.cloudfoundry.promregator.lite.cfaccessor.SimpleCFAccessorImpl
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.coroutineScope
+import org.cloudfoundry.client.v2.routes.Route
+import org.cloudfoundry.promregator.lite.config.CloudFoundryConfig
+import org.cloudfoundry.promregator.lite.config.PromregatorConfiguration.Companion.DEFAULT_ID
+import org.springframework.stereotype.Component
+import java.util.*
+import java.util.concurrent.atomic.AtomicBoolean
+import org.cloudfoundry.promregator.lite.config.Target
+import kotlin.text.RegexOption.IGNORE_CASE
+
+
+data class OrgResponse(
+        val api: String,
+        val id: String,
+        val name: String
+)
+
+data class SpaceResponse(
+        val org: OrgResponse,
+        val spaceName: String,
+        val spaceId: String,
+        val domains: List<DomainResponse>,
+)
+
+data class SpaceSummaryAppResponse(
+        val space: SpaceResponse,
+        val appName: String,
+        val appSummary: AppSummary
+)
+
+data class AppSummary(
+        val instances: Int,
+        val urls: List<String>,
+        val routes: List<Route>,
+)
+
+data class DomainResponse(
+        val orgId: String,
+        val domainId: String,
+        val isInternal: Boolean,
+)
+
+data class AppResponse(
+        val appId: String,
+        val appName: String,
+        val space: SpaceResponse,
+        val isScrappable: Boolean,
+        val annotationMetricsPath: String? = null,
+        val originalTarget: Target? = null,
+        val spaceAppSummary: AppSummary? = null,
+) {
+    val protocol: String
+        get() = originalTarget?.protocol ?: "https"
+    val metricsPath: String
+        get() = annotationMetricsPath ?: originalTarget?.path ?: "/metrics"
+}
+
+private val log = mu.KotlinLogging.logger {  }
+
+@Component
+class SimpleTargetResolver(
+        private val cfAccessor: SimpleCFAccessorImpl,
+        private val cf: CloudFoundryConfig
+) {
+    suspend fun resolveTargets(configTargetsToMatch: List<Target>): List<AppResponse> = coroutineScope {
+        val configTargets = configTargetsToMatch.ifEmpty { listOf(Target()) }
+
+        val orgs = resolveOrgs(configTargets)
+        val domains = orgs.resolveDomains()
+        val spaces = orgs.resolveSpaces(configTargets, domains)
+
+        val apps = if (cfAccessor.isV3Enabled) {
```

Comment:
I look forward to expanding on the v3 usage now that it is so much easier to reason about. Now the logic to only gather the space summaries when v3 is enabled is much simpler to follow.

Reply:
When v3 is available, this path now uses all v3 endpoints! I need to fix the logic that calculates internal routes though.